File: evals/refs/karat.go
package main

import (
	"bufio"
	"os"
	"strconv"
)

// Karatsuba multiplication on little-endian base-256 byte arrays.
// Pure []byte digit ops (no math/big, no encoding/binary).

const cutoff = 32

// trim removes leading (most-significant) zero digits, keeping at least one.
func trim(a []byte) []byte {
	n := len(a)
	for n > 1 && a[n-1] == 0 {
		n--
	}
	return a[:n]
}

// add returns a+b as a new little-endian byte slice.
func add(a, b []byte) []byte {
	n := len(a)
	if len(b) > n {
		n = len(b)
	}
	out := make([]byte, n+1)
	carry := 0
	for i := 0; i < n; i++ {
		s := carry
		if i < len(a) {
			s += int(a[i])
		}
		if i < len(b) {
			s += int(b[i])
		}
		out[i] = byte(s & 0xFF)
		carry = s >> 8
	}
	out[n] = byte(carry)
	return trim(out)
}

// sub returns a-b assuming a >= b, trimmed.
func sub(a, b []byte) []byte {
	out := make([]byte, len(a))
	borrow := 0
	for i := 0; i < len(a); i++ {
		d := int(a[i]) - borrow
		if i < len(b) {
			d -= int(b[i])
		}
		if d < 0 {
			d += 256
			borrow = 1
		} else {
			borrow = 0
		}
		out[i] = byte(d)
	}
	return trim(out)
}

// school does O(n*m) schoolbook multiplication.
func school(a, b []byte) []byte {
	out := make([]int, len(a)+len(b))
	for i := 0; i < len(a); i++ {
		x := int(a[i])
		if x == 0 {
			continue
		}
		carry := 0
		for j := 0; j < len(b); j++ {
			s := out[i+j] + x*int(b[j]) + carry
			out[i+j] = s & 0xFF
			carry = s >> 8
		}
		k := i + len(b)
		for carry != 0 {
			s := out[k] + carry
			out[k] = s & 0xFF
			carry = s >> 8
			k++
		}
	}
	res := make([]byte, len(out))
	for i, v := range out {
		res[i] = byte(v)
	}
	return trim(res)
}

func karat(a, b []byte) []byte {
	if len(a) <= cutoff || len(b) <= cutoff {
		return school(a, b)
	}
	m := len(a)
	if len(b) < m {
		m = len(b)
	}
	m /= 2
	a0, a1 := a[:m], a[m:]
	b0, b1 := b[:m], b[m:]
	z0 := karat(a0, b0)
	z2 := karat(a1, b1)
	z1 := sub(sub(karat(add(a0, a1), add(b0, b1)), z0), z2)

	out := make([]int, len(a)+len(b)+4)
	for i, v := range z0 {
		out[i] += int(v)
	}
	for i, v := range z1 {
		out[i+m] += int(v)
	}
	for i, v := range z2 {
		out[i+2*m] += int(v)
	}
	carry := 0
	for i := range out {
		s := out[i] + carry
		out[i] = s & 0xFF
		carry = s >> 8
	}
	res := make([]byte, len(out))
	for i, v := range out {
		res[i] = byte(v)
	}
	return trim(res)
}

func main() {
	rd := bufio.NewReaderSize(os.Stdin, 1<<20)
	w := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer w.Flush()

	readInt := func() int {
		n := 0
		c, err := rd.ReadByte()
		for err == nil && (c == ' ' || c == '\n' || c == '\r' || c == '\t') {
			c, err = rd.ReadByte()
		}
		for err == nil && c >= '0' && c <= '9' {
			n = n*10 + int(c-'0')
			c, err = rd.ReadByte()
		}
		return n
	}

	t := readInt()
	for c := 0; c < t; c++ {
		na := readInt()
		nb := readInt()
		a := make([]byte, na)
		b := make([]byte, nb)
		for i := 0; i < na; i++ {
			a[i] = byte(readInt())
		}
		for i := 0; i < nb; i++ {
			b[i] = byte(readInt())
		}
		p := karat(trim(a), trim(b))
		for i, v := range p {
			if i > 0 {
				w.WriteByte(' ')
			}
			w.WriteString(strconv.Itoa(int(v)))
		}
		w.WriteByte('\n')
	}
}

File: evals/refs/karat.cpp
// Karatsuba multiplication on little-endian base-256 byte arrays.
// std::vector<uint8_t> digit ops; no __int128 / gmp / boost.
// g++ -O2 -std=c++17 -Wall -Wextra
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>

using std::size_t;
using Digits = std::vector<uint8_t>;

static const size_t CUTOFF = 32;

static void trim(Digits &n) {
    while (n.size() > 1 && n.back() == 0) n.pop_back();
}

// Views into a backing vector avoid copies on recursion.
struct View {
    const uint8_t *p;
    size_t n;
};

static Digits add(View a, View b) {
    size_t n = a.n > b.n ? a.n : b.n;
    Digits out(n + 1, 0);
    int carry = 0;
    for (size_t i = 0; i < n; i++) {
        int s = carry;
        if (i < a.n) s += a.p[i];
        if (i < b.n) s += b.p[i];
        out[i] = static_cast<uint8_t>(s & 0xFF);
        carry = s >> 8;
    }
    out[n] = static_cast<uint8_t>(carry);
    trim(out);
    return out;
}

// a - b, assuming a >= b.
static Digits sub(View a, View b) {
    Digits out(a.n, 0);
    int borrow = 0;
    for (size_t i = 0; i < a.n; i++) {
        int d = static_cast<int>(a.p[i]) - borrow;
        if (i < b.n) d -= b.p[i];
        if (d < 0) { d += 256; borrow = 1; } else { borrow = 0; }
        out[i] = static_cast<uint8_t>(d);
    }
    trim(out);
    return out;
}

static Digits school(View a, View b) {
    std::vector<uint32_t> acc(a.n + b.n, 0);
    for (size_t i = 0; i < a.n; i++) {
        uint32_t x = a.p[i];
        if (x == 0) continue;
        uint32_t carry = 0;
        for (size_t j = 0; j < b.n; j++) {
            uint32_t s = acc[i + j] + x * static_cast<uint32_t>(b.p[j]) + carry;
            acc[i + j] = s & 0xFF;
            carry = s >> 8;
        }
        size_t k = i + b.n;
        while (carry) {
            uint32_t s = acc[k] + carry;
            acc[k] = s & 0xFF;
            carry = s >> 8;
            k++;
        }
    }
    Digits out(acc.size());
    for (size_t i = 0; i < acc.size(); i++) out[i] = static_cast<uint8_t>(acc[i]);
    trim(out);
    return out;
}

static Digits karat(View a, View b) {
    if (a.n <= CUTOFF || b.n <= CUTOFF) return school(a, b);

    size_t m = (a.n < b.n ? a.n : b.n) / 2;
    View a0{a.p, m}, a1{a.p + m, a.n - m};
    View b0{b.p, m}, b1{b.p + m, b.n - m};

    Digits z0 = karat(a0, b0);
    Digits z2 = karat(a1, b1);
    Digits sa = add(a0, a1);
    Digits sb = add(b0, b1);
    Digits zm = karat(View{sa.data(), sa.size()}, View{sb.data(), sb.size()});
    Digits t1 = sub(View{zm.data(), zm.size()}, View{z0.data(), z0.size()});
    Digits z1 = sub(View{t1.data(), t1.size()}, View{z2.data(), z2.size()});

    size_t total = a.n + b.n + 4;
    std::vector<uint32_t> acc(total, 0);
    for (size_t i = 0; i < z0.size(); i++) acc[i] += z0[i];
    for (size_t i = 0; i < z1.size(); i++) acc[i + m] += z1[i];
    for (size_t i = 0; i < z2.size(); i++) acc[i + 2 * m] += z2[i];

    uint32_t carry = 0;
    for (size_t i = 0; i < total; i++) {
        uint32_t s = acc[i] + carry;
        acc[i] = s & 0xFF;
        carry = s >> 8;
    }
    Digits out(total);
    for (size_t i = 0; i < total; i++) out[i] = static_cast<uint8_t>(acc[i]);
    trim(out);
    return out;
}

int main() {
    // Slurp stdin.
    std::string in;
    {
        char buf[1 << 16];
        size_t r;
        while ((r = fread(buf, 1, sizeof(buf), stdin)) > 0) in.append(buf, r);
    }
    size_t pos = 0;
    auto read_int = [&]() -> long {
        while (pos < in.size() && (in[pos] < '0' || in[pos] > '9')) pos++;
        long n = 0;
        while (pos < in.size() && in[pos] >= '0' && in[pos] <= '9') {
            n = n * 10 + (in[pos] - '0');
            pos++;
        }
        return n;
    };

    std::string out;
    out.reserve(1 << 20);

    long t = read_int();
    for (long c = 0; c < t; c++) {
        long na = read_int();
        long nb = read_int();
        Digits a(static_cast<size_t>(na));
        Digits b(static_cast<size_t>(nb));
        for (long i = 0; i < na; i++) a[i] = static_cast<uint8_t>(read_int());
        for (long i = 0; i < nb; i++) b[i] = static_cast<uint8_t>(read_int());
        trim(a);
        trim(b);

        Digits p = karat(View{a.data(), a.size()}, View{b.data(), b.size()});
        for (size_t i = 0; i < p.size(); i++) {
            if (i > 0) out.push_back(' ');
            out += std::to_string(p[i]);
        }
        out.push_back('\n');
    }
    fwrite(out.data(), 1, out.size(), stdout);
    return 0;
}
